### ISSUE TITLE
Photos not showing in popup

### ISSUE THREAD

**Issue by HerbFargus** [tomchadwin/qgis2web#661]
Qgis2Web Version: 3.1.1
QGIS Version: 3.0.2-Girona
Platform: Windows 10

I apologise in advance if this is a user error, if so disregard. 

In my case I have a field with a list of URLS and I want the URLS to show up as a photo in the popup instead of an HTML Link.

Here is the layer:
[test.zip](https://github.com/tomchadwin/qgis2web/files/2208289/test.zip)

QGIS 3 is a bit different and has an attachment widget with options to specify an `image` or `web view` whereas qgis2.X just has a photo widget.

![image](https://user-images.githubusercontent.com/10035308/42922650-cc7b64d2-8ade-11e8-83ef-47cb84118860.png)

If I set it to web view:
![image](https://user-images.githubusercontent.com/10035308/42922670-e34c14d6-8ade-11e8-894b-78b90ebd7bab.png)

Using Identify in QGIS yields:

![image](https://user-images.githubusercontent.com/10035308/42922708-1e3c5a2e-8adf-11e8-8b95-236a01d97748.png)

If I export the layer in leaflet or openlayers I just get a clickable link:

![image](https://user-images.githubusercontent.com/10035308/42922744-5339588a-8adf-11e8-849f-0f78b947a492.png)

If I set it to image:

![image](https://user-images.githubusercontent.com/10035308/42922682-f8f91bf8-8ade-11e8-87be-b306bef7b591.png)

basically the same, just a clickable link:

![image](https://user-images.githubusercontent.com/10035308/42922744-5339588a-8adf-11e8-849f-0f78b947a492.png)

I'm not sure if this is due to the changes in QGIS3, or if I am misunderstanding how it is supposed to work (maybe I'm missing a setting?). I can manually generate HTML tables if I need to but I presume that automation is the main intent for the plugin.

Let me know if you need anything else, I'm happy to provide any logs or additional testing as needed.

**Comment by tomchadwin** [comment]
Others have reported this (https://twitter.com/SylMaithya/status/1012380722148139009), but there doesn't seem to be an issue here, so thanks for opening this.

My problem is that I can't investigate because I'm getting a QGIS crash when I set the widget type to Attachment. I've opened a QGIS issue:

https://issues.qgis.org/issues/19118

Until the QGIS issue is fixed, I can't really make any progress.

**Comment by magerlin** [comment]
Same issue here - and no problems setting widget type to Attachment
Win 7 64bit, QGIS 3.2.2
QGIS2Web 3.2.0

**Comment by magerlin** [comment]
Please note: Same problem is when using QGIS version 2.18.23 and setting widget type to Photo (or Web, my photos are on another web server)

**Comment by ben-chmark** [comment]
Hallo, 

I had the same problem with my photos and i fixed it.

Openlayers:

Step 1:
Chose the widget type "Attachment" for the photos.
and write the photoname into the column like you did - redwood.jpg

To view the photos in the browser you have to go into the created js-files after exporting from qgis.

Step 2:

Go to the folder layers and open the layers.js and change

lyr_UmsetzungStecken2017_14.set('fieldImages', {'OBJECTID': 'Hidden', 'HR': 'TextEdit', 'ID': 'TextEdit', 'Shape_Leng': 'Hidden', 'Photo': 'ExternalResource', });

 to

lyr_UmsetzungStecken2017_14.set('fieldImages', {'OBJECTID': 'Hidden', 'HR': 'TextEdit', 'ID': 'TextEdit', 'Shape_Leng': 'Hidden', 'Photo': 'Attachment', });

Step 3:

Go to the folder resources and open the qgis2web.js file

search for the popup 

and change 

if (layer.get('fieldImages')[currentFeatureKeys[i]] != "Photo") {
  popupField += (clusterFeature.get(currentFeatureKeys[i]) != null ? Autolinker.link(String(clusterFeature.get(currentFeatureKeys[i]))) + '</td>' : '');
                                } 
to

if (layer.get('fieldImages')[currentFeatureKeys[i]] != "Attachment") {
                                    popupField += (clusterFeature.get(currentFeatureKeys[i]) != null ? Autolinker.link(String(clusterFeature.get(currentFeatureKeys[i]))) + '</td>' : '');
                                }

after this this, you can view the photos in the browser



Leaflet:

with the Leaflet_framework, you only have to open the index-file and search for the popupContent and change

 <tr>\
             <td colspan="2"><strong>Photo</strong><br />' + (feature.properties['Photo'] !== null ? Autolinker.link(String(feature.properties['Photo'])) : '') + '</td>\
                    </tr>\

to

 <tr>\
      <td colspan="2"><img src="images/' + (feature.properties['Photo'] !== null ? Autolinker.link(String(feature.properties['Photo'])) : '') + '" /></td>\
                    </tr>\


Ben

**Comment by ben-chmark** [comment]
sorry 

Leaflet:
change 


<tr>\
  <td colspan="2"><strong>Photo</strong><br />' + (feature.properties['Photo'] !== null ? Autolinker.link(String(feature.properties['Photo'])) : '') + '</td>\
                    </tr>\

to

<td colspan="2"><img src="images/' + (feature.properties['Photo'] !== null ? Autolinker.link(String(feature.properties['Photo'])) : '') + '" /></td>\

**Comment by jukos** [comment]
@BenDienstl 

>  Leaflet:
> change
> \ Photo
> ' + (feature.properties['Photo'] !== null ? Autolinker.link(String(feature.properties['Photo'])) : '') + '\ \
> 
> to
> 
> \ 

to.....
please write a text, I can not see the image....
Thank you and greetings

**Comment by jukos** [comment]
Hello Tom,
any news?
You have created this ticket
https://issues.qgis.org/issues/19118
before 11 month......
Can I/we hope for the next version? Do you know anything?

To set attachments with images/videos for qgis2web, I must use QGIS 2.18.....but I would like work with QGIS 3.


Thank you for response. 

Jürgen

**Comment by aQuafr3sh** [comment]
> 
> 
> sorry
> 
> Leaflet:
> change
> \ **Photo**
> ' + (feature.properties['Photo'] !== null ? Autolinker.link(String(feature.properties['Photo'])) : '') + '\ \
> 
> to
> 
> ![](images/' + (feature.properties['Photo'] !== null ? Autolinker.link(String(feature.properties['Photo'])) : '') + ')
> \

@BenDienstl can you please explain the "to" part? It appears to be linking to an image of some sort. When I hover over the link it shows the same information as the "from" part.

Thanks!

**Comment by apenzin** [comment]
> 
> 
> Hallo,
> 
> I had the same problem with my photos and i fixed it.
> 
> Openlayers:
> 
> Step 1:
> Chose the widget type "Attachment" for the photos.
> and write the photoname into the column like you did - redwood.jpg
> 
> To view the photos in the browser you have to go into the created js-files after exporting from qgis.
> 
> Step 2:
> 
> Go to the folder layers and open the layers.js and change
> 
> lyr_UmsetzungStecken2017_14.set('fieldImages', {'OBJECTID': 'Hidden', 'HR': 'TextEdit', 'ID': 'TextEdit', 'Shape_Leng': 'Hidden', 'Photo': 'ExternalResource', });
> 
> to
> 
> lyr_UmsetzungStecken2017_14.set('fieldImages', {'OBJECTID': 'Hidden', 'HR': 'TextEdit', 'ID': 'TextEdit', 'Shape_Leng': 'Hidden', 'Photo': 'Attachment', });
> 
> Step 3:
> 
> Go to the folder resources and open the qgis2web.js file
> 
> search for the popup
> 
> and change
> 
> if (layer.get('fieldImages')[currentFeatureKeys[i]] != "Photo") {
> popupField += (clusterFeature.get(currentFeatureKeys[i]) != null ? Autolinker.link(String(clusterFeature.get(currentFeatureKeys[i]))) + '' : '');
> }
> to
> 
> if (layer.get('fieldImages')[currentFeatureKeys[i]] != "Attachment") {
> popupField += (clusterFeature.get(currentFeatureKeys[i]) != null ? Autolinker.link(String(clusterFeature.get(currentFeatureKeys[i]))) + '' : '');
> }
> 
> after this this, you can view the photos in the browser
> 
> Leaflet:
> 
> with the Leaflet_framework, you only have to open the index-file and search for the popupContent and change
> \ **Photo**
> ' + (feature.properties['Photo'] !== null ? Autolinker.link(String(feature.properties['Photo'])) : '') + '\ \
> 
> to
> \
> 
> ![](images/' + (feature.properties['Photo'] !== null ? Autolinker.link(String(feature.properties['Photo'])) : '') + ')
> \ \
> 
> Ben

Dear Ben,

Worked perfectly. I would like to know the code to link the thumbnail images of the popup with the image display at its original size and resolution in another tab in the browser. Is it possible?

Amadeu

**Comment by jukos** [comment]
Hello,
it looks like that the issue is solved:
https://github.com/qgis/QGIS/issues/26948

> Yes, this should bring images in popups back to life in qgis2web. However, no doubt there will be issues at the qgis2web end by now, as I've not looked at that code for a long time. Now that we have the QGIS bug fixed, and once you have a working version of QGIS with that fix, let's continue the discussion back on the qgis2web repo.
(Tom Chadwin)

**@tomchadwin** 
To testing the improvement (+ possible adjustments within qgis2web)  you must wait for the next QGIS version, or it is for you possible to test it earlier than 
3.10.0 | 3.4.13 |   | 2019-10-25 **?**

Greetings, Jürgen

**Comment by tomchadwin** [comment]
@jukos You can install the nightly, and test with that - that should have the upstream QGIS fix in it.

**Comment by jukos** [comment]
Could you give me the link to "nightly"?
Thank you

**Comment by tomchadwin** [comment]
You can install the nightly via the OSGeo installer: https://qgis.org/en/site/forusers/alldownloads.html#windows

CAVEAT: I can take no responsibility for the installation of QGIS versions on your machine - please ensure you read up and understand how to do what you need.

**Comment by jukos** [comment]
Hmm. I have tried the weekly version 

QGIS-OSGeo4W-3.9.0-38-Setup-x86.exe | 24-Sep-2019 15:14 | 411M |  
-- | -- | -- | --

without success. Only this version I could test. Problems with the other one.
I do not know, if the improvement is included in this weekly version...

Maybe another one can do any test. If not, I am waiting to the regular qgis version.

**Comment by jukos** [comment]
Tom Chadwin:
https://github.com/qgis/QGIS/issues/26948:
> Now that we have the QGIS bug fixed, and once you have a working version of QGIS with that fix, let's continue the discussion back on the qgis2web repo.

(only) my test on the actual weekly 3.9 version: no button crash
I hope you can use this version for your analysis to get back the images.....
I hope also you have time the next days/weeks.

Greetings

**Comment by jukos** [comment]
@ Tom Chadwin
It is possible to have the qgis2web photo / image function for qgis3 for the next qgis version 

LTR/PR | 3.10.0 | 3.4.13 |   | 2019-10-25
-- | -- | -- | -- | --

What do you think? 

Thank you for answer and greetings
